ktlint-ruleset-standard/src/main/kotlin/com/github/shyiko/ktlint/ruleset/standard/IndentationRule.kt

```diff
@@ -2,81 +2,312 @@ package com.github.shyiko.ktlint.ruleset.standard
                 }
+            } else {
+                children
+                    .toList()
+                    .reversed()
+                    .forEach { stack.push(it) }
             }
-            if (node.textContains('\t')) {
-                val text = node.getText()
-                emit(node.startOffset + text.indexOf('\t'), "Unexpected Tab character(s)", true)
-                if (autoCorrect) {
-                    (node as LeafPsiElement).rawReplaceWithText(text.replace("\t", " ".repeat(indentSize)))
+        }
+    }
+
+    private fun handleNewline(
+        node: PsiWhiteSpace,
+        emit: (offset: Int, errorMessage: String, canBeAutoCorrected: Boolean) -> Unit,
+        autoCorrect: Boolean,
+        normalIndentationCount: Int,
+        continuationIndentationCount: Int
+    ) {
+        val lines = node.text.split("\n")
+        if (
+            lines.size > 1 &&
+            // parameter list wrapping enforced by ParameterListWrappingRule
+            !node.isPartOf(KtParameterList::class)
+        ) {
+            var offset = node.startOffset + lines.first().length + 1
+
+            val inBetweenLines = lines.tail().take(lines.count() - 2)
+
+            inBetweenLines.forEach { indent ->
+                if (indent != "") {
+                    emit(
+                        offset,
+                        "Unexpected indentation (${indent.length}) (it should be 0)",
+                        true
+                    )
                 }
+
+                offset += indent.length + 1
+            }
+
+            val indent = lines.last()
+
+            val totalIndentationSize = normalIndentationCount * indentSize +
+                continuationIndentationCount * continuationIndentSize
+
+            if (indent.length != totalIndentationSize) {
+                emit(
+                    offset,
+                    "Unexpected indentation (${indent.length}) (it should be $totalIndentationSize)",
+                    true
+                )
+            }
+
+            if (autoCorrect) {
+                val fixedString = node.text
+                    .replace("""\n.*\n""", """\n\n""")
+                    .replaceAfterLast('\n', " ".repeat(totalIndentationSize))
+                (node as LeafPsiElement).rawReplaceWithText(fixedString)
             }
         }
     }
 
-    private fun gcd(a: Int, b: Int): Int = when {
-        a > b -> gcd(a - b, b)
-        a < b -> gcd(a, b - a)
-        else -> a
+    private fun handleTab(
+        node: ASTNode,
+        emit: (offset: Int, errorMessage: String, canBeAutoCorrected: Boolean) -> Unit,
+        autoCorrect: Boolean
+    ) {
+        emit(
+            node.startOffset + node.text.indexOf('\t'),
+            "Unexpected Tab character(s)",
+            true
+        )
+
+        if (autoCorrect) {
+            val tabToSpaceString = node.text.replace("\t", " ".repeat(indentSize))
+            (node as LeafPsiElement).rawReplaceWithText(tabToSpaceString)
```

Comment:
I'm wondering whether it would be safe to use `as` here. Is there any other cases that `node` might be some other kinds than `LeafPsiElement`?